crates/phactory/src/contracts/geolocation.rs

```diff
@@ -33,28 +41,49 @@ pub enum Error {
 
 #[derive(Encode, Decode, Debug, Clone)]
 pub enum Request {
-    GetGeolocationInfo { account: AccountId },
-    GetAvailableCityName {},
-    GetCityDistribution { city_name: String },
-    GetCityDistributionCount { city_name: String },
+    GetGeocoding { account: AccountId },
+    GetAvailableRegionName,
+    GetAccountsInRegion { region_name: String },
+    GetAccountCountInRegion { region_name: String },
 }
 
 #[derive(Encode, Decode, Debug, Clone)]
 pub enum Response {
-    GetGeolocationInfo { geolocation_info: CoordinateInfo },
-    GetAvailableCityName { city_names: Vec<String> },
-    GetCityDistribution { workers: Vec<AccountId> },
-    GetCityDistributionCount { count: u32 },
+    GetGeocoding { geocoding: Geocoding },
+    GetAvailableRegionName { region_names: Vec<String> },
+    GetAccountsInRegion { workers: Vec<AccountId> },
+    GetAccountCountInRegion { count: u32 },
     Error(String),
 }
 
 impl Geolocation {
     pub fn new() -> Self {
         Geolocation {
-            geolocation_info: BTreeMap::new(),
-            city_distribution: BTreeMap::new(),
+            geo_data: BTreeMap::new(),
+            region_map: BTreeMap::new(),
         }
     }
+
+    pub fn purge(&mut self, current_blocknum: &chain::BlockNumber) {
+        // purging expired region map
+        for (k, v) in &self.geo_data {
```

Comment:
I think this pieces of code can be optimized. We can first scan the full worker list to filter out the expired data (O(n)). Then we scan the reserve index and filter out all the workers if they are not in the map (again, O(n)).

The current implementation has a worst complexity of O(n^2)

Reply:
Ah, I think probably an easier way is to just remove the reverse index at all. Then we leave an API to build the aggregation on-the-fly.

Or we can leave the reverse index, but build it every time at on_block_end. It's an O(n) operation anyway. In the current code, when processing the command, it's also O(n) in the worse case, or O(a) in the average case (a is the average number of workers per city).

Reply:
You are right, I was overthinking this issue, it should be much simpler than current implementation.